[core] Enhance the spinlock type, used extensively in the fiber implementation

kjcamann opened the pull request:
This PR contains debugging enhancements to the existing `spinlock_t` type, which is renamed to `monad_spinlock_t`. This is an isolated change, because the spinlock was not currently used by any other code. To simplify the fiber implementation, there is no fancy lock-free programming in the subsequent PRs; the all major objects (the fibers, the synchronization primitives, and the scheduler) are simple implementations covered by a single spinlock. The locks are held for very short periods, which can be verified by enabling the locking statistics option. The other feature this adds is source location tracking for the location where the lock was last taken, so it adds a formal source location object for C, similar to `std::source_location`, and a C++ conversion function between the two.

Commit:
[core][spinlock] Enhance the spinlock type; spinlocks are used
extensively in the fiber implementation

This adds source location tracking for the location where the lock
is taken, so it adds a formal source location object for C, similar
to std::source_location, and a C++ conversion function between the
two

It makes a few other changes:

  - cpu_relax() is renamed to monad_spinloop_hint(), which is a
    more faithful representation of what it is, and adds some
    documentation explaining the behavior of this intrinsic at the
    CPU level
  - cpu_relax.h was renamed to spinloop.h
  - The loop itself was changed to pause after every failed
    iteration, rather than after a fixed-size adaptive spin to check
    if the lock is free. This improves the included lock contention
    benchmark by 50%. This is not surprising, as the intention
    of PAUSE is to explicitly give a spin-loop hint to the CPU,
    to defeat whatever bad speculative execution optimizations
    it might do internally. The old implementation was doing a lot
    of spin-waiting without providing this hint, so likely the CPU
    does something better now.

ned14 reviewed (commented):
Firstly, obligatory https://www.reddit.com/r/programming/comments/ejyr3r/mutexes_are_faster_than_spinlocks/

Secondly, can't say I like source location for this at all. It's nearly useless. We already have a stack backtrace capture and print functionality in Core. I'd suggest that instead.

Finally, me personally, I'd never use spinlocks around any code which isn't wait free. I'd even go so far that if a spin count takes too long in debug, we abort the process and force the dev to use a mutex. That's probably considered excessive, but I've seen many a codebase ruined by too many spinlocks because "they're faster", and they're only faster if the code they protect is extremely short and well bounded. Mutexes already have a short spin count in them, and the ones in libgcc are "magical" in that the compiler does clever tricks with them.

Commit:
[core][spinlock] placement new requires #include <new> with libc++

Commit:
[core][spinlock] address some of Niall's comments

Also increase the number of threads in the micro-benchmark; this
more faithfully tests the effect of backoff when contention is high

ned14 reviewed (dismissed):
Minor nitpicks only

Commit:
[core][spinlock] Address Niall's final comments

kjcamann committed:
[core][spinlock] spinlock tests cannot fight with other parallel ctests

The test deliberately has very high contention, and will run far too
slowly if the threads cannot actually spin (e.g., because the system
is under heavy load)

kjcamann commented:
Closing a defunct PR
